Sources/TensorFlow/Optimizers/LearningRates.swift

```diff
@@ -0,0 +1,348 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Learning rate schedule that takes a learning rate as input, along with the current training
+/// step and returns a modified learning rate (e.g., decayed).
+public protocol LearningRateSchedule {
+    associatedtype Scalar: FloatingPoint
+
+    /// Returns the transformed value of `learningRate` for the specified training step.
+    ///
+    /// - Parameters:
+    ///   - step: Training step.
+    ///   - learningRate: Learning rate value to transform according to this schedule.
+    func callAsFunction(step: UInt64, learningRate: Scalar) -> Scalar
+}
+
+extension LearningRateSchedule {
+    /// Returns a new learning rate schedule which is defined as the composition of this schedule
+    /// with the provided one. More specifically, the new schedule will transform the provided
+    /// learning rates as follows:
+    /// `self(step: step, learningRate: other(step: step, learningRate: learningRate))`.
+    public func composed<Schedule: LearningRateSchedule>(
+        with other: Schedule
+    ) -> ComposedLearningRateSchedule<Self, Schedule> {
+        ComposedLearningRateSchedule(schedule1: self, schedule2: other)
+    }
+}
+
+/// Composed learning rate schedule.
+///
+/// This schedule transforms the provided learning rates as follows:
+/// ```
+/// schedule1(step: step, learningRate: schedule2(step: step, learningRate: learningRate))
+/// ```
+public struct ComposedLearningRateSchedule<
+    Schedule1: LearningRateSchedule,
+    Schedule2: LearningRateSchedule
+>: LearningRateSchedule where Schedule1.Scalar == Schedule2.Scalar {
+    public typealias Scalar = Schedule1.Scalar
+
+    public let schedule1: Schedule1
```

Comment:
IMHO and perhaps a nitpick, but to me using 1 and 2 on the names is a bit misleading as schedule2 is performed before schedule1 and these numbers could be confused to denote the ordering of the schedules.

Reply:
That's a good point. The composed schedule is not there anymore so this confusion is now avoided. :)